UPSTREAM: <carry>: use node.Spec.ProviderID for identifying nodegroups

frobware opened the pull request:
This change switches from using `node.Name` to node.Spec.ProviderID when building a `cloudprovider.NodeGroup`. Using `node.Name` has been a historical mistake.

There have been several suggestions for adding the provider ID to the machine object at the cluster-api level (https://github.com/kubernetes-sigs/cluster-api/pull/565) and by an actuator implementation (https://github.com/openshift/cluster-api-provider-aws/pull/86). For the moment we can make this mapping without either of those PRs assuming there's the "machine" annotation on the node object. This annotation is currently added by the nodelink-controller (https://github.com/openshift/machine-api-operator).

Switching `cloudprovider.NodeGroupForNode()` to indexing on `node.Spec.ProviderID` and also returning provider ID values in `cloudprovider.Nodes()` means we no longer experience the case where the nodegroup/node becomes unregistered, which ultimately leads to the autoscaler to stop both scale up and down operations as it deems the state of the cluster to be unhealthy.

We may want to revisit how this is done in the longer term because it does assume that the nodelink-controller is running on the cluster. But this change requires no additional changes from the cluster-api (so no revendoring) nor does it require a change to the actuators that we use (AWS, libvirt).

frobware commented:
/cc @derekwaynecarr @ingvagabund @bison @vikaschoudhary16

ingvagabund commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -21,57 +21,61 @@ import (
-	objs, err := machineIndexer.ByIndex(nodeNameIndexKey, node.Name)
+func (c *machineController) findMachine(id string) (*v1alpha1.Machine, error) {
+	store := c.machineInformer.Informer().GetStore()
+	item, exists, err := store.GetByKey(id)
```
Given the `store` is used only here, it's more transparent to do `c.machineInformer.Informer().GetStore().GetByKey(id)` here.

frobware replied:
Done.

ingvagabund commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -92,45 +96,145 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
 
-// run starts shared informers and waits for the shared informer cache
-// to synchronize.
+// Run starts shared informers and waits for the informer cache to
```
`s/Run/run`

frobware replied:
Done.

ingvagabund commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -92,45 +96,145 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
+// machineSet. For each machine in the set a DeepCopy() of the object
+// is returned.
+func (c *machineController) MachinesInMachineSet(machineSet *v1alpha1.MachineSet) ([]*v1alpha1.Machine, error) {
+	machines, err := c.machineInformer.Lister().Machines(machineSet.Namespace).List(labels.Everything())
```
This is fine with not so many machines. Though, with hundreds of machines in the same namespace (which would be insine but still) this can be slow. What's the benefit of listing all instead of listing all machines that has labels matched by the machineset's label selector?

frobware replied:
Done.

ingvagabund commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go:
```diff
@@ -68,8 +70,14 @@ func (p *provider) nodeNames(machineSet *v1alpha1.MachineSet) ([]string, error)
+		}
+
+		if node != nil {
+			nodeNames = append(nodeNames, node.Spec.ProviderID)
```
Why do you append the ProviderID into the list of node names?

frobware replied:
The variable name is badly named now.

frobware replied:
Renamed.

frobware commented:
@ingvagabund addressed your issues. @tnozicka had some feedback that I should also be checking UUID values which may be better to do in a follow-up PR as it's important that we switch to indexing via node.spec.ProviderID ASAP. Thoughts? @derekwaynecarr

bison commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -92,45 +94,146 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
+	if err != nil {
+		return nil, err
+	}
+	if len(objs) != 1 {
```
Should this return an error if the length is greater than one? However unlikely, it would be good to know.

frobware replied:
Done.

bison commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -92,45 +94,146 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
 	}
 
-	return nil
+	machineName, found := node.Annotations["machine"]
```
Seems like this should probably try the newer `cluster.k8s.io/machine` annotation first, then fallback to `machine`.

frobware replied:
Done. Thanks for the link.

frobware commented:
> @tnozicka had some feedback that I should also be checking UUID values

This is now done too.

frobware commented:
/retest

derekwaynecarr commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -92,45 +93,151 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
+
+	machineName, found := node.Annotations["cluster.k8s.io/machine"]
+	if !found {
+		machineName, found = node.Annotations["machine"]
```
q: we seem to fall back on this always currently, is there a separate pr that updates our annotation usage?
